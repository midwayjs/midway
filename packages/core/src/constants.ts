/**
 * 静态参数
 *
 */
export const KEYS = {
  OBJECTS_ELEMENT: 'objects',
  OBJECT_ELEMENT: 'object',
  IMPORT_ELEMENT: 'import',
  PROPERTY_ELEMENT: 'property',
  LIST_ELEMENT: 'list',
  MAP_ELEMENT: 'map',
  ENTRY_ELEMENT: 'entry',
  VALUE_ELEMENT: 'value',
  PROPS_ELEMENT: 'props',
  PROP_ELEMENT: 'prop',
  SET_ELEMENT: 'set',
  CONSTRUCTOR_ARG_ELEMENT: 'constructor-arg',
  REF_ELEMENT: 'ref',
  JSON_ELEMENT: 'json',
  CONFIGURATION_ELEMENT: 'configuration',

  ID_ATTRIBUTE: 'id',
  PATH_ATTRIBUTE: 'path',
  DIRECT_ATTRIBUTE: 'direct',
  AUTOWIRE_ATTRIBUTE: 'autowire',
  ASYNC_ATTRIBUTE: 'async',
  NAME_ATTRIBUTE: 'name',
  REF_ATTRIBUTE: 'ref',
  KEY_ATTRIBUTE: 'key',
  VALUE_ATTRIBUTE: 'value',
  TYPE_ATTRIBUTE: 'type',
  EXTERNAL_ATTRIBUTE: 'external',
  OBJECT_ATTRIBUTE: 'object',
  RESOURCE_ATTRIBUTE: 'resource',
  SCOPE_ATTRIBUTE: 'scope',

  ASPECT_ELEMENT: 'aspect',
  AROUND_ELEMENT: 'around',
  EXPRESSION_ATTRIBUTE: 'expression',
  EXECUTE_ATTRIBUTE: 'execute',
};

export const FUNCTION_INJECT_KEY = 'midway:function_inject_key';
export const MIDWAY_LOGGER_WRITEABLE_DIR = 'MIDWAY_LOGGER_WRITEABLE_DIR';
export const REQUEST_CTX_KEY = 'ctx';
export const REQUEST_OBJ_CTX_KEY = '_req_ctx';
export const HTTP_SERVER_KEY = '_midway_http_server';
export const REQUEST_CTX_LOGGER_CACHE_KEY = '_midway_ctx_logger_cache';

export const ASYNC_CONTEXT_KEY = Symbol('ASYNC_CONTEXT_KEY');
export const ASYNC_CONTEXT_MANAGER_KEY = 'MIDWAY_ASYNC_CONTEXT_MANAGER_KEY';

export const DEFAULT_PATTERN = [
  '**/**.ts',
  '**/**.js',
  '**/**.mts',
  '**/**.mjs',
  '**/**.cts',
  '**/**.cjs',
];

export const IGNORE_PATTERN = ['**/**.d.ts', '**/**.d.mts', '**/**.d.cts'];
