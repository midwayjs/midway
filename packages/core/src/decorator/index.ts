// common
export * from './common/provide';
export * from './common/inject';
export * from './common/pipeline';
export * from './common/aspect';
export * from './common/autoload';
export * from './common/configuration';
export * from './common/objectDef';
export * from './common/framework';
export * from './common/filter';
export * from './common/middleware';
export * from './common/guard';
export * from './common/pipe';
export * from './common/mock';

// faas
export * from './faas/serverlessTrigger';

// web
export * from './web/controller';
export * from './web/paramMapping';
export * from './web/requestMapping';
export * from './web/response';

// other
export * from './constant';
export * from './decoratorManager';

// microservice
export * from './microservice/consumer';
export * from './microservice/provider';
export * from './microservice/rabbitmqListener';
export * from './microservice/kafkaListener';

// rpc
export * from './rpc/hsf';

// task
export * from './task/queue';
export * from './task/task';
export * from './task/taskLocal';
export * from './task/schedule';

// ws
export * from './ws/webSocketController';
export * from './ws/webSocketEvent';
