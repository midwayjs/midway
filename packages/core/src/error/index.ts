export * from './base';
export { HttpStatus, httpError } from './http';
export * from './framework';
