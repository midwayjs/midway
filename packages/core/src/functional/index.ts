export { defineConfiguration } from './configuration';
export * from './hooks';
export * from './interface';
