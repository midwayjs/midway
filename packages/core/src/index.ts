export {
  ObjectIdentifier,
  ObjectDefinitionOptions,
  IManagedInstance,
  ScopeEnum,
  saveClassMetadata,
  attachClassMetadata,
  getClassMetadata,
  saveMethodDataToClass,
  attachMethodDataToClass,
  getMethodDataFromClass,
  listMethodDataFromClass,
  saveMethodMetadata,
  attachMethodMetadata,
  getMethodMetadata,
  savePropertyDataToClass,
  attachPropertyDataToClass,
  getPropertyDataFromClass,
  listPropertyDataFromClass,
  savePropertyMetadata,
  attachPropertyMetadata,
  getPropertyMetadata,
  savePreloadModule,
  listPreloadModule,
  saveModule,
  listModule,
  resetModule,
  clearAllModule,
  getParamNames,
  getProviderId,
  getObjectDefinition,
  classNamed,
  generateProvideId
} from '@midwayjs/decorator';
export * from './interface';
export { ContainerLoader } from './loader';
export { MidwayContainer } from './context/midwayContainer';
export { MidwayRequestContainer } from './context/requestContainer';
export { BaseFramework } from './baseFramework';
export * from './context/providerWrapper';
export * from './common/constants';
export {
  safelyGet,
  safeRequire,
  parsePrefix,
} from './common/util';
export * from './features';
export * from './util/webRouterParam';
export { plainToClass, classToPlain } from 'class-transformer';
