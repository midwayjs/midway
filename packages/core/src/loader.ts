import * as path from 'path';
import { MidwayContainer } from './context/midwayContainer';
import {
  CONFIGURATION_KEY,
  getProviderId,
  listModule,
} from '@midwayjs/decorator';
import { ILifeCycle, IMidwayContainer } from './interface';

function buildLoadDir(baseDir, dir) {
  if (!path.isAbsolute(dir)) {
    return path.join(baseDir, dir);
  }
  return dir;
}

/**
 * @deprecated
 */
export class ContainerLoader {
  baseDir;
  applicationContext: MidwayContainer;
  isTsMode;
  preloadModules;
  disableConflictCheck: boolean;

  constructor({
    baseDir,
    isTsMode = true,
    preloadModules = [],
    disableConflictCheck = true,
  }) {
    this.baseDir = baseDir;
    this.isTsMode = isTsMode;
    this.preloadModules = preloadModules;
    this.disableConflictCheck = disableConflictCheck;
  }

  initialize() {
    this.applicationContext = new MidwayContainer(this.baseDir, undefined);
    this.applicationContext.disableConflictCheck = this.disableConflictCheck;
    this.applicationContext.registerObject('baseDir', this.baseDir);
    this.applicationContext.registerObject('isTsMode', this.isTsMode);
  }

  getApplicationContext(): IMidwayContainer {
    return this.applicationContext;
  }

  registerHook(hookKey, hookHandler) {
    this.applicationContext.registerDataHandler(hookKey, hookHandler);
  }

  loadDirectory(
    loadOpts: {
      baseDir?: string;
      loadDir?: string[];
      disableAutoLoad?: boolean;
      pattern?: string | string[];
      ignore?: string | string[];
    } = {}
  ) {
    if (!this.isTsMode && loadOpts.disableAutoLoad === undefined) {
      // disable auto load in js mode by default
      loadOpts.disableAutoLoad = true;
    }

    if (loadOpts.disableAutoLoad) return;

    // use baseDir in parameter first
    const baseDir = loadOpts.baseDir || this.baseDir;
    const defaultLoadDir = this.isTsMode ? [baseDir] : [];
    this.applicationContext.load({
      loadDir: (loadOpts.loadDir || defaultLoadDir).map(dir => {
        return buildLoadDir(baseDir, dir);
      }),
      pattern: loadOpts.pattern,
      ignore: loadOpts.ignore,
    });

    if (this.preloadModules && this.preloadModules.length) {
      for (const preloadModule of this.preloadModules) {
        this.applicationContext.bindClass(preloadModule);
      }
    }
  }

  async refresh() {
    await this.applicationContext.ready();
    await this.loadLifeCycles();
  }

  async stop() {
    await this.applicationContext.stop();
  }

  async loadLifeCycles() {
    // agent 不加载生命周期
    const cycles: Array<{ target: any; namespace: string }> = listModule(
      CONFIGURATION_KEY
    );
    for (const cycle of cycles) {
      const providerId = getProviderId(cycle.target);
      const inst = await this.getApplicationContext().getAsync<ILifeCycle>(
        providerId
      );
      if (typeof inst.onReady === 'function') {
        /**
         * 让组件能正确获取到 bind 之后 registerObject 的对象有三个方法
         * 1、在 load 之后修改 bind，不太可行
         * 2、每次 getAsync 的时候，去掉 namespace，同时还要查找当前全局的变量，性能差
         * 3、一般只会在 onReady 的地方执行 registerObject（否则没有全局的意义），这个取巧的办法就是 onReady 传入一个代理，其中绑定当前的 namespace
         */
        await inst.onReady(
          new Proxy(this.getApplicationContext(), {
            get: function (target, prop, receiver) {
              if (prop === 'getCurrentNamespace' && cycle.namespace) {
                return () => {
                  return cycle.namespace;
                };
              }
              return Reflect.get(target, prop, receiver);
            },
          })
        );
      }
    }
  }
}
