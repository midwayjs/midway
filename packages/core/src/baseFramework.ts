import {
  IConfigurationOptions,
  IMidwayApplication,
  IMidwayBootstrapOptions,
  IMidwayContainer,
  IMidwayFramework,
  MidwayFrameworkType,
  MidwayProcessTypeEnum,
} from './interface';
import { ContainerLoader } from './';
import { APPLICATION_KEY } from '@midwayjs/decorator';

export abstract class BaseFramework<
  APP extends IMidwayApplication,
  T extends IConfigurationOptions
> implements IMidwayFramework<APP, T> {
  protected isTsMode = true;
  protected baseDir: string;
  protected appDir: string;
  protected containerLoader: ContainerLoader;
  public configurationOptions: T;
  public app: APP;

  public configure(options: T): BaseFramework<APP, T> {
    this.configurationOptions = options;
    return this;
  }

  public async initialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {
    this.baseDir = options.baseDir;
    this.appDir = options.appDir;

    this.containerLoader = new ContainerLoader({
      baseDir: this.baseDir,
      isTsMode: this.isTsMode,
      preloadModules: options.preloadModules || [],
    });

    /**
     * initialize containerLoader and initialize ioc container instance
     */
    await this.beforeInitialize(options);
    this.containerLoader.initialize();

    /**
     * load directory and bind files to ioc container
     */
    await this.beforeDirectoryLoad(options);
    const applicationContext = this.containerLoader.getApplicationContext();
    applicationContext.registerObject('baseDir', this.baseDir);
    applicationContext.registerObject('appDir', this.appDir);
    // 如果没有关闭autoLoad 则进行load
    this.containerLoader.loadDirectory(options);

    // register app
    this.containerLoader.registerHook(APPLICATION_KEY, () => {
      return this.getApplication();
    });

    await this.afterDirectoryLoad(options);

    if (!this.app.getApplicationContext) {
      this.app = this.defineApplicationProperties(this.app);
    }

    /**
     * start to load configuration and lifeCycle
     */
    await this.containerLoader.refresh();
    await this.afterInitialize(options);
  }

  public getApplicationContext(): IMidwayContainer {
    return this.containerLoader.getApplicationContext();
  }

  public getConfiguration(key?: string): any {
    return this.getApplicationContext()
      .getConfigService()
      .getConfiguration(key);
  }

  public getCurrentEnvironment() {
    return this.getApplicationContext()
      .getEnvironmentService()
      .getCurrentEnvironment();
  }

  public abstract getFrameworkType(): MidwayFrameworkType;

  public abstract getApplication(): APP;

  public abstract run(): Promise<void>;

  public async stop(): Promise<void> {
    await this.beforeStop();
    await this.containerLoader.stop();
  }

  protected defineApplicationProperties(app: APP): APP {
    return Object.assign(app, {
      getBaseDir: () => {
        return this.baseDir;
      },

      getAppDir: () => {
        return this.appDir;
      },

      getEnv: () => {
        return this.getApplicationContext()
          .getEnvironmentService()
          .getCurrentEnvironment();
      },

      getConfig: (key?: string) => {
        return this.getApplicationContext()
          .getConfigService()
          .getConfiguration(key);
      },

      getFrameworkType: () => {
        return this.getFrameworkType();
      },

      getProcessType: () => {
        return MidwayProcessTypeEnum.APPLICATION;
      },
    });
  }

  protected async beforeStop(): Promise<void> {}

  protected async beforeInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async beforeDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async afterDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected abstract async afterInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void>;
}
