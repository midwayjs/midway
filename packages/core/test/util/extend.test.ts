import { extend } from '../../src';

const str = 'me a test';
const integer = 10;
const arr = [1, 'what', new Date(81, 8, 4)];
const date = new Date(81, 4, 13);

const Foo = function () {};

const obj = {
  str: str,
  integer: integer,
  arr: arr,
  date: date,
  constructor: 'fake',
  isPrototypeOf: 'not a function',
  foo: new Foo()
};

const deep = {
  ori: obj,
  layer: {
    integer: 10,
    str: 'str',
    date: new Date(84, 5, 12),
    arr: [101, 'dude', new Date(82, 10, 4)],
    deep: {
      str: obj.str,
      integer: integer,
      arr: obj.arr,
      date: new Date(81, 7, 4)
    }
  }
};

describe('util/extend.test.ts', function () {
  it('missing arguments', function () {
    expect(extend(undefined, { a: 1 })).toStrictEqual({ a: 1 });
    expect(extend({ a: 1 })).toStrictEqual({ a: 1 });
    expect(extend(true, undefined, { a: 1 })).toStrictEqual({ a: 1 });
    expect(extend(true, { a: 1 })).toStrictEqual({ a: 1 });
    expect(extend()).toStrictEqual({});
  });

  it('merge string with string', function () {
    var ori = 'what u gonna say';
    var target = extend(ori, str);
    var expectedTarget = {
      0: 'm',
      1: 'e',
      2: ' ',
      3: 'a',
      4: ' ',
      5: 't',
      6: 'e',
      7: 's',
      8: 't'
    };

    expect(target).toEqual(expectedTarget);
  });

  it('merge string with number', function () {
    var ori = 'what u gonna say';
    var target = extend(ori, 10);

    expect(target).toEqual({});
  });

  it('merge string with array', function () {
    var ori = 'what u gonna say';
    var target = extend(ori, arr);

    expect(arr).toEqual([1, 'what', new Date(81, 8, 4)]);
    expect(target).toEqual({
      0: 1,
      1: 'what',
      2: new Date(81, 8, 4)
    });
  });

  it('merge string with date', function () {
    var ori = 'what u gonna say';
    var target = extend(ori, date);

    var testDate = new Date(81, 4, 13);
    expect(date).toEqual(testDate);
    expect(target).toEqual({});
  });

  it('merge string with obj', function () {
    var ori = 'what u gonna say';
    var target = extend(ori, obj);

    var testObj = {
      str: 'me a test',
      integer: 10,
      arr: [1, 'what', new Date(81, 8, 4)],
      date: new Date(81, 4, 13),
      constructor: 'fake',
      isPrototypeOf: 'not a function',
      foo: new Foo()
    };
    expect(obj).toEqual(testObj);
    expect(target).toEqual(testObj);
  });

  it('merge number with string', function () {
    var ori = 20;
    var target = extend(ori, str);

    expect(target).toEqual({
      0: 'm',
      1: 'e',
      2: ' ',
      3: 'a',
      4: ' ',
      5: 't',
      6: 'e',
      7: 's',
      8: 't'
    });
  });

  it('merge number with number', function () {
    expect(extend(20, 10)).toEqual({});
  });

  it('merge number with array', function () {
    var target = extend(20, arr);

    expect(arr).toEqual([1, 'what', new Date(81, 8, 4)]);
    expect(target).toEqual({
      0: 1,
      1: 'what',
      2: new Date(81, 8, 4)
    });
  });

  it('merge number with date', function () {
    var target = extend(20, date);
    var testDate = new Date(81, 4, 13);

    expect(date).toEqual(testDate);
    expect(target).toEqual({});
  });

  it('merge number with object', function () {
    var target = extend(20, obj);
    var testObj = {
      str: 'me a test',
      integer: 10,
      arr: [1, 'what', new Date(81, 8, 4)],
      date: new Date(81, 4, 13),
      constructor: 'fake',
      isPrototypeOf: 'not a function',
      foo: new Foo()
    };

    expect(obj).toEqual(testObj);
    expect(target).toEqual(testObj);
  });

  it('merge array with string', function () {
    var ori = [1, 2, 3, 4, 5, 6];
    var target = extend(ori, str);

    expect(ori).toEqual(str.split(''));
    expect(target).toEqual([
      "m",
      "e",
      " ",
      "a",
      " ",
      "t",
      "e",
      "s",
      "t"
    ]);
  });

  it('merge array with number', function () {
    var ori = [1, 2, 3, 4, 5, 6];
    var target = extend(ori, 10);

    expect(ori).toEqual([1, 2, 3, 4, 5, 6]);
    expect(target).toEqual(ori);
  });

  it('merge array with array', function () {
    var ori = [1, 2, 3, 4, 5, 6];
    var target = extend(ori, arr);
    var testDate = new Date(81, 8, 4);
    var expectedTarget = [1, 'what', testDate, 4, 5, 6];

    expect(ori).toEqual(expectedTarget);
    expect(arr).toEqual([1, 'what', testDate]);
    expect(target).toEqual(expectedTarget);
  });

  it('merge array with date', function () {
    var ori = [1, 2, 3, 4, 5, 6];
    var target = extend(ori, date);
    var testDate = new Date(81, 4, 13);
    var testArray = [1, 2, 3, 4, 5, 6];

    expect(ori).toEqual(testArray);
    expect(date).toEqual(testDate);
    expect(target).toEqual(testArray);
  });

  it('merge array with object', function () {
    var ori = [1, 2, 3, 4, 5, 6];
    var target = extend(ori, obj);
    var testObject = {
      str: 'me a test',
      integer: 10,
      arr: [1, 'what', new Date(81, 8, 4)],
      date: new Date(81, 4, 13),
      constructor: 'fake',
      isPrototypeOf: 'not a function',
      foo: new Foo()
    };

    expect(obj).toStrictEqual(testObject);
    expect(target.length).toEqual(6);
    expect(target.integer).toEqual(obj.integer);
    expect(target.arr).toEqual(obj.arr);
    expect(target.date).toEqual(obj.date);
  });

  it('merge date with string', function () {
    var ori = new Date(81, 9, 20);
    var target = extend(ori, str);

    expect(target).toEqual(ori);
  });

  it('merge date with number', function () {
    var ori = new Date(81, 9, 20);
    var target = extend(ori, 10);

    expect(target).toEqual(ori);
  });

  it('merge date with array', function () {
    var ori = new Date(81, 9, 20);
    var target = extend(ori, arr);
    var testDate = new Date(81, 9, 20);
    var testArray = [1, 'what', new Date(81, 8, 4)];

    expect(ori).toEqual(testDate);
    expect(arr).toEqual(testArray);
    expect(target).toEqual(testDate);
  });

  it('merge date with date', function () {
    var ori = new Date(81, 9, 20);
    var target = extend(ori, date);

    expect(target).toEqual(ori);
  });

  it('merge date with object', function () {
    var ori = new Date(81, 9, 20);
    var target = extend(ori, obj);

    expect(target).toEqual(ori);
  });

  it('merge object with string', function () {
    var testDate = new Date(81, 7, 26);
    var ori = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: testDate
    };
    var target = extend(ori, str);
    var testObj = {
      0: 'm',
      1: 'e',
      2: ' ',
      3: 'a',
      4: ' ',
      5: 't',
      6: 'e',
      7: 's',
      8: 't',
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: testDate
    };

    expect(ori).toEqual(testObj);
    expect(target).toEqual(testObj);
  });

  it('merge object with number', function () {
    var ori = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26)
    };
    var testObject = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26)
    };
    var target = extend(ori, 10);
    expect(ori).toEqual(testObject);
    expect(target).toEqual(testObject);
  });

  it('merge object with array', function () {
    var ori = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26)
    };
    var target = extend(ori, arr);
    var testObject = {
      0: 1,
      1: 'what',
      2: new Date(81, 8, 4),
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26)
    };

    expect(ori).toEqual(testObject);
    expect(arr).toEqual([1, 'what', testObject[2]]);
    expect(target).toEqual(testObject);

  });

  it('merge object with date', function () {
    var ori = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26)
    };
    var target = extend(ori, date);
    var testObject = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26)
    };

    expect(ori).toEqual(testObject);
    expect(date).toEqual(new Date(81, 4, 13));
    expect(target).toEqual(testObject);

  });

  it('merge object with object', function () {
    var ori = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26),
      foo: 'bar'
    };
    var target = extend(ori, obj);
    var expectedObj = {
      str: 'me a test',
      integer: 10,
      arr: [1, 'what', new Date(81, 8, 4)],
      date: new Date(81, 4, 13),
      constructor: 'fake',
      isPrototypeOf: 'not a function',
      foo: new Foo()
    };
    var expectedTarget = {
      str: 'me a test',
      integer: 10,
      arr: [1, 'what', new Date(81, 8, 4)],
      date: new Date(81, 4, 13),
      constructor: 'fake',
      isPrototypeOf: 'not a function',
      foo: new Foo()
    };

    expect(obj).toEqual(expectedObj);
    expect(ori).toEqual(expectedTarget);
    expect(target).toEqual(expectedTarget);

  });

  it('deep clone', function () {
    var ori = {
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26),
      layer: { deep: { integer: 42 } }
    };
    var target = extend(true, ori, deep);

    expect(ori).toEqual({
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26),
      ori: {
        str: 'me a test',
        integer: 10,
        arr: [1, 'what', new Date(81, 8, 4)],
        date: new Date(81, 4, 13),
        constructor: 'fake',
        isPrototypeOf: 'not a function',
        foo: new Foo()
      },
      layer: {
        integer: 10,
        str: 'str',
        date: new Date(84, 5, 12),
        arr: [101, 'dude', new Date(82, 10, 4)],
        deep: {
          str: 'me a test',
          integer: 10,
          arr: [1, 'what', new Date(81, 8, 4)],
          date: new Date(81, 7, 4)
        }
      }
    });
    expect(deep).toEqual({
      ori: {
        str: 'me a test',
        integer: 10,
        arr: [1, 'what', new Date(81, 8, 4)],
        date: new Date(81, 4, 13),
        constructor: 'fake',
        isPrototypeOf: 'not a function',
        foo: new Foo()
      },
      layer: {
        integer: 10,
        str: 'str',
        date: new Date(84, 5, 12),
        arr: [101, 'dude', new Date(82, 10, 4)],
        deep: {
          str: 'me a test',
          integer: 10,
          arr: [1, 'what', new Date(81, 8, 4)],
          date: new Date(81, 7, 4)
        }
      }
    });
    expect(target).toEqual({
      str: 'no shit',
      integer: 76,
      arr: [1, 2, 3, 4],
      date: new Date(81, 7, 26),
      ori: {
        str: 'me a test',
        integer: 10,
        arr: [1, 'what', new Date(81, 8, 4)],
        date: new Date(81, 4, 13),
        constructor: 'fake',
        isPrototypeOf: 'not a function',
        foo: new Foo()
      },
      layer: {
        integer: 10,
        str: 'str',
        date: new Date(84, 5, 12),
        arr: [101, 'dude', new Date(82, 10, 4)],
        deep: {
          str: 'me a test',
          integer: 10,
          arr: [1, 'what', new Date(81, 8, 4)],
          date: new Date(81, 7, 4)
        }
      }
    });

    target.layer.deep = 339;
    expect(deep).toEqual({
      ori: {
        str: 'me a test',
        integer: 10,
        arr: [1, 'what', new Date(81, 8, 4)],
        date: new Date(81, 4, 13),
        constructor: 'fake',
        isPrototypeOf: 'not a function',
        foo: new Foo()
      },
      layer: {
        integer: 10,
        str: 'str',
        date: new Date(84, 5, 12),
        arr: [101, 'dude', new Date(82, 10, 4)],
        deep: {
          str: 'me a test',
          integer: 10,
          arr: [1, 'what', new Date(81, 8, 4)],
          date: new Date(81, 7, 4)
        }
      }
    });
    // ----- NEVER USE EXTEND WITH THE ABOVE SITUATION ------------------------------

  });

  it('deep clone; arrays are override', function () {
    var defaults = { arr: [1, 2, 3] };
    var override = { arr: ['x'] };
    var expectedTarget = { arr: ['x'] };

    var target = extend(true, defaults, override);

    expect(target).toEqual(expectedTarget);

  });

  it('deep clone === false; objects merged normally', function () {
    var defaults = { a: 1 };
    var override = { a: 2 };
    var target = extend(false, defaults, override);
    expect(target).toEqual(override);

  });

  it('pass in null; should create a valid object', function () {
    var override = { a: 1 };
    var target = extend(null, override);
    expect(target).toEqual(override);

  });

  it.skip('works without Array.isArray', function () {
    var savedIsArray = Array.isArray;
    (Array as any).isArray = false; // don't delete, to preserve enumerability
    var target = [];
    var source = [1, [2], { 3: true }];
    expect(
      extend(true, target, source)).toEqual(
      [1, [2], { 3: true }],
    );
    Array.isArray = savedIsArray;
  });

  it('fix __proto__ copy', function () {
    const r = extend(true, {}, JSON.parse('{"__proto__": {"polluted": "yes"}}'));
    expect(
      JSON.stringify(r)).toEqual(
      '{}'
    );
    expect(
      ('' as any).polluted).toEqual(
      undefined,
    );
  });

  it('should test null merge', function () {
    let r = extend(true, {
      a: 1
    }, {a: null});
    expect(r.a).toBe(null);

    r = extend(true, {
      a: undefined
    }, {a: null});
    expect(r.a).toBe(null);

    r = extend(true, {
      a: null
    }, {a: undefined});
    expect(r.a).toBe(null);
  });

});
