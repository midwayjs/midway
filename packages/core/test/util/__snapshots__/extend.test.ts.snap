// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/extend.test.ts merge date with date 1`] = `1981-10-19T16:00:00.000Z`;

exports[`util/extend.test.ts merge date with date 2`] = `1981-10-19T16:00:00.000Z`;

exports[`util/extend.test.ts merge date with number 1`] = `1981-10-19T16:00:00.000Z`;

exports[`util/extend.test.ts merge date with number 2`] = `1981-10-19T16:00:00.000Z`;

exports[`util/extend.test.ts merge date with object 1`] = `
Object {
  "arr": Array [
    1,
    "what",
    1981-09-03T16:00:00.000Z,
  ],
  "constructor": "fake",
  "date": 1981-05-12T16:00:00.000Z,
  "foo": Foo {},
  "integer": 10,
  "isPrototypeOf": "not a function",
  "str": "me a test",
}
`;

exports[`util/extend.test.ts merge date with object 2`] = `1981-10-19T16:00:00.000Z`;

exports[`util/extend.test.ts merge date with object 3`] = `1981-10-19T16:00:00.000Z`;

exports[`util/extend.test.ts merge date with string 1`] = `1981-10-19T16:00:00.000Z`;
